src/Advection/weno5.jl

```diff
@@ -0,0 +1,274 @@
+#####
+##### Weighted Essentially Non-Oscillatory (WENO) scheme of order 5
+#####
+
+struct WENO5 <: AbstractAdvectionScheme end
+
+#####
+##### ENO interpolants of size 3
+#####
+
+@inline px₀(i, j, k, f) = @inbounds + 1/3 * f[i-1, j, k] + 5/6 * f[i,   j, k] -  1/6 * f[i+1, j, k]
+@inline px₁(i, j, k, f) = @inbounds - 1/6 * f[i-2, j, k] + 5/6 * f[i-1, j, k] +  1/3 * f[i,   j, k]
+@inline px₂(i, j, k, f) = @inbounds + 1/3 * f[i-3, j, k] - 7/6 * f[i-2, j, k] + 11/6 * f[i-1, j, k]
+
+@inline py₀(i, j, k, f) = @inbounds + 1/3 * f[i, j-1, k] + 5/6 * f[i, j,   k] -  1/6 * f[i, j+1, k]
+@inline py₁(i, j, k, f) = @inbounds - 1/6 * f[i, j-2, k] + 5/6 * f[i, j-1, k] +  1/3 * f[i, j  , k]
+@inline py₂(i, j, k, f) = @inbounds + 1/3 * f[i, j-3, k] - 7/6 * f[i, j-2, k] + 11/6 * f[i, j-1, k]
+
+@inline pz₀(i, j, k, f) = @inbounds + 1/3 * f[i, j, k-1] + 5/6 * f[i, j,   k] -  1/6 * f[i, j, k+1]
+@inline pz₁(i, j, k, f) = @inbounds - 1/6 * f[i, j, k-2] + 5/6 * f[i, j, k-1] +  1/3 * f[i, j,   k]
+@inline pz₂(i, j, k, f) = @inbounds + 1/3 * f[i, j, k-3] - 7/6 * f[i, j, k-2] + 11/6 * f[i, j, k-1]
+
+#####
+##### Jiang & Shu (1996) WENO smoothness indicators
+#####
+
+@inline βx₀(i, j, k, f) = @inbounds 13/12 * (f[i-1, j, k] - 2f[i,   j, k] + f[i+1, j, k])^2 + 1/4 * (3f[i-1, j, k] - 4f[i,   j, k] +  f[i+1, j, k])^2
+@inline βx₁(i, j, k, f) = @inbounds 13/12 * (f[i-2, j, k] - 2f[i-1, j, k] + f[i,   j, k])^2 + 1/4 * ( f[i-2, j, k]                 -  f[i,   j, k])^2
+@inline βx₂(i, j, k, f) = @inbounds 13/12 * (f[i-3, j, k] - 2f[i-2, j, k] + f[i-1, j, k])^2 + 1/4 * ( f[i-3, j, k] - 4f[i-2, j, k] + 3f[i-1, j, k])^2
+
+@inline βy₀(i, j, k, f) = @inbounds 13/12 * (f[i, j-1, k] - 2f[i, j,   k] + f[i, j+1, k])^2 + 1/4 * (3f[i, j-1, k] - 4f[i,   j, k] +  f[i, j+1, k])^2
+@inline βy₁(i, j, k, f) = @inbounds 13/12 * (f[i, j-2, k] - 2f[i, j-1, k] + f[i, j,   k])^2 + 1/4 * ( f[i, j-2, k]                 -  f[i,   j, k])^2
+@inline βy₂(i, j, k, f) = @inbounds 13/12 * (f[i, j-3, k] - 2f[i, j-2, k] + f[i, j-1, k])^2 + 1/4 * ( f[i, j-3, k] - 4f[i, j-2, k] + 3f[i, j-1, k])^2
+
+@inline βz₀(i, j, k, f) = @inbounds 13/12 * (f[i, j, k-1] - 2f[i, j,   k] + f[i, j, k+1])^2 + 1/4 * (3f[i, j, k-1] - 4f[i, j,   k] +  f[i, j, k+1])^2
+@inline βz₁(i, j, k, f) = @inbounds 13/12 * (f[i, j, k-2] - 2f[i, j, k-1] + f[i, j,   k])^2 + 1/4 * ( f[i, j, k-2]                 -  f[i, j,   k])^2
+@inline βz₂(i, j, k, f) = @inbounds 13/12 * (f[i, j, k-3] - 2f[i, j, k-2] + f[i, j, k-1])^2 + 1/4 * ( f[i, j, k-3] - 4f[i, j, k-2] + 3f[i, j, k-1])^2
+
+#####
+##### WENO-5 optimal weights
+#####
+
+const C3₀ = 3/10
+const C3₁ = 3/5
+const C3₂ = 1/10
+
+#####
+##### WENO-5 raw weights
+#####
+
+const ε = 1e-6
```

Comment:
This parameter could belong to the `WENO` struct --- unless it really cant be changed. But if its a rounding error thing, it might depend on precision?

Reply:
Hmmm yeah they are properties of the `WENO{N}` struct but for `WENO5` I just hard-coded them since I'm thinking of deprecating `WENO5` once `WENO{N}` is fast. Plus, I guess we tested for convergence and periodic advection + thermal bubble with `ε = 1e-6` so I guess I don't see an immediate reason  to change its value (also kinda lazy about refactoring).